linux-mod-r1.eclass: add support for sys-kernel/dkms

Nowa-Ammerlaan opened the pull request:
This can use more polishing, but lets discuss the concept first now that this proof of principle is working.

The idea here is to optionally install external kernel modules with support for `sys-kernel-dkms`. This is useful since dkms makes it possible to easily install a module for multiple kernel versions. Via an init service it is also possible to compile modules on the fly if they are found to be missing at boot (thus avoid failures if you accidentally forgot to `emerge @module-rebuild`).  It is a tool that is used in many other distributions to manage external kernel modules, and thus will be familiar to many new users. Upstream already has tests for Gentoo and we have had this packaged for a while now, the only part that is lacking is the integration with portage managed ebuilds, this is what we address here.

With regards to the integration with portage, this is what is respected:
- User and ebuild patches, since we install the `S` after all of these have been applied
- The users compiler and flags preferences, since this is part of the `emakeargs` inserted into `MAKE[0]`.
- `USE=strip` via `STRIP[.]`
What is not respected (currently):
- The users `MAKEOPTS` (controlled via dkms configuration) , we could add the `MAKEOPTS` to the `make` command, but this would override the dkms configuration which may be confusing.
- `USE=modules-sign` and `USE=modules-compress` (also controlled via dkms configuration), you can point dkms to the `MODULES_SIGN_KEY` and `MODULES_SIGN_CERT` manually though, and also configure the compression command.

Drawbacks: 
- Ebuilds that do not use `linux-mod-r1`'s compile and install functions would have to provide their own `dkms.conf`, otherwise the flag effectively does nothing.
- We need a `pkg_prerm` function that we did not need before.

Tested and found to be working with `virtualbox-modules`, `nvidia-drivers` and `droidcam`.

Let me know what you think of this.

CC @ionenwks  

---

Please check all the boxes that apply:

- [ ] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [ ] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [ ] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [ ] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

ionenwks commented on eclass/linux-mod-r1.eclass:
```diff
@@ -1353,4 +1438,4 @@ _modules_update_depmod() {
 fi
 
-EXPORT_FUNCTIONS pkg_setup src_compile src_install pkg_postinst
+EXPORT_FUNCTIONS pkg_setup src_compile src_install pkg_postinst pkg_prerm
```
I don't want to introduce new exports at this point without either doing linux-mod-r2 or doing it on EAPI bump, this eclass is meant to be stable at this point.

ionenwks commented on eclass/linux-mod-r1.eclass:
```diff
@@ -421,8 +423,16 @@ linux-mod-r1_src_compile() {
 	local -a emakeargs=( "${MODULES_MAKEARGS[@]}" )
 	[[ ${modargs@a} == *a* ]] && emakeargs+=( "${modargs[@]}" )
 
+	if use dkms; then
```
Note that using linux-mod-r1_src_compile and src_install is optional (see eclassdoc's examples) and cannot be relied on, e.g. zfs-kmod does not use it for one and all that logic will be unused. It does require that packages call `modules_post_process` in src_install though.

The eclass also allows linux-mod-r1_src_compile to be called multiple times (noted in its description) which I guess would overwrite this dkms.conf.

ionenwks replied:
Ah, nevermind I was re-reading your post and you already had that listed in drawbacks for the first bit.

ionenwks reviewed (commented):
Can't say I particularly like having another path to support, nor compilation being handled by something else than portage.

While I can see some use cases, I'm not convinced that we really need to support this.

Haven't looked at changes closely or tried it, but for a few extra comments...

ionenwks commented:
>This is useful since dkms makes it possible to easily install a module for multiple kernel versions.

If we really wanted this (not that I imagine that many users need it), technically this could be implemented in the eclasses directly (aka have it loop for multiple kernels) -- albeit I think it'd be rather messy unless we rewrite linux-info and do a linux-mod-r2 to go with it (possibly with some forearch_kernel wrappers and such).

Not to say should actually work on this, just something to keep on the table if we rework this eventually.

It's non-intuititve but a user can do this manually by re-emerging the modules for each kernels given portage does not delete them.

Nowa-Ammerlaan commented:
Maybe this would be cleaner as a separate eclass that somehow works on top off `linux-mod-r1.eclass`. This should mostly resolve the problem that we would otherwise create for ebuilds that use only the `modules_post_process` function.

thesamesam commented:
I prefer the idea of a separate eclass not least to avoid adding a `USE` which may or may not work properly (and may lead to us accidentally "encouraging" dkms use) on all packages. Better to use it in a handful of cases or advertise it as a way for people to more easily package modules in overlays if they're intimidated.

Nowa-Ammerlaan commented:
Alright, scraped my first attempt and started fresh. Now I added a whole new `dkms.eclass` containing some helper functions, I also significantly improved the logic to generate a `dkms.conf` and added `pkg_config` functions for convenience. I think this new approach is a lot cleaner.

Using the new eclass I added support for `USE=dkms` to several ebuilds. This is working well on my end, though in some packages some tweaking is involved to modify/move existing `dkms.conf` files. Later I will expand this to more packages installing kernel modules.

Let me know if you have any comments. @ionenwks  I made a very small modification to `linux-mod-r1.eclass` which makes life a lot simpler for me in `dkms.eclass`.

eli-schwartz commented on app-emulation/virtualbox-guest-additions/virtualbox-guest-additions-7.1.6-r1.ebuild:
```diff
@@ -203,10 +203,18 @@ src_compile() {
+	fi
 }
 
 src_install() {
 	linux-mod-r1_src_install
+	dkms_src_install
```
So if I understand this correctly, packages with USE=dkms will *both* build modules as has been standard up to now on Gentoo, directly in the package, and *also* install dkms sources that are intended for building the module which was already produced (and others).

This feels wrong, it should be either one or the other. If you want to use dkms, use dkms. If you want to use `emerge @module-rebuild`, use that.

I do understand the appeal of dkms. Especially, it was exceedingly useful on a pure binary-packages distribution, and could be useful with a binhost too I suppose. e.g. you may be able to avoid patching sources and painfully running eautoreconf (looking at you zfs lmao) and arguably has stronger legal standing in the case of modules that violate the GPL (looking at you ***again***, zfs!)

ionenwks commented:
Like I said in my initial "review", I'm more against the idea itself than whichever implementation is used for it, esp. the bit about having 2 ways to do the same thing, making things harder to maintain and control and just bringing more potential problems and things to consider.

I don't feel like that we really need this.

Nowa-Ammerlaan commented:
> So if I understand this correctly, packages with USE=dkms will _both_ build modules as has been standard up to now on Gentoo, directly in the package, and _also_ install dkms sources that are intended for building the module which was already produced (and others).

Yes, this is a good point. I had initially thought of this as an extension upon what `linux-mod-r1.eclass` was already doing, but indeed now we are doing things twice. An if/else structure would make more sense here.

> I'll reflect on it a bit more and also see what other's comments are. I think I could be persuaded as to its value with more description, a promise of good docs which won't lead to confused users, and also some solution to the IMO ugliness of the snippets in the consumer ebuilds.

> Like I said in my initial "review", I'm more against the idea itself than whichever implementation is used for it, esp. the bit about having 2 ways to do the same thing, making things harder to maintain and control and just bringing more potential problems and things to consider.

I understand your hesitation. I also had to warm up to this idea after initially discarding it for precisely the concerns you are now raising (two paths to achieve similar ends and the possible confusion and added maintenance load as a result of this). However, after thinking about it more I am now convinced this is worth the effort (I would not have written the code if I was not convinced of this after all).

- Dynamically building missing modules at boot is I think the "killer feature" here. This is something we cannot reasonably achieve using the package manager. I think this brings much needed reliability and "just works experience" to Gentoo's external kernel module support. Gentoo is a distro that facilitates the parallel installation of multiple different kernel versions/branches more than most other distros do. However, currently we can only target one single kernel for building an external kernel module (yes this limitation can be circumvented but its not easy and not pretty), this leads to a messy situation when upgrading kernel modules:
  -  Old versions are left un-upgraded on kernels that were not targeted (resulting in confusing mismatches between what portage/userspace reports and what the kernel is actually using).
     - We have several imperfect workarounds for this in place (zfs, zfs-kmod, dist-kernel_reinstall_initramfs), but none of that reliably resolves the problem.
  - There is no way to remove a kernel module once installed (e.g. leading to confused users who switch from nvidia to nouveau)
  - If you reboot into a kernel that is missing a external networking module, you might have a difficult time resolving the situation due to being unable to access the network.
  - If you're missing a GPU module, granted you can fix this, but it is unpleasant as it breaks the desktop completely.
  - While you're resolving the situation of a missing module, you'll often have to run a command like this: `eselect kernel set x && FEATURES="-getbinpkg -buildpkg" emerge @modules-rebuild --ignore-default-opts` because all this currently integrates poorly with binary packages. We know that we can workaround the problem like this, but for new users this command is non-trivial.
  - Upgrading a distribution kernel works fine, downgrading is non-trivial (and portage is slow in resolving the slot conflicts).
-  We can very easily resolve all of this by letting go of insisting that the package manager must resolve these problems. And instead simply installing the sources for these modules and having a little bash script and init service (i.e. DKMS) take care of building the kernel modules.
- We can do this while still respecting the users C/CXX/CFLAGS/CXXFLAGS/LDFLAGS/MAKEOPTS/PATCHES etc, so we do not lose any user customization in this process.
- DKMS is the industry standard solution to the problem of managing external kernel modules, as such many upstreams already support and test this path.
- Gentoo being source-based does not really change any of this, DKMS is not exclusively useful for binary distributions.
- Everything is opt-in, so the simple path of installing a kernel module for the eselected kernel version remains possible (for e.g. cross-compiling situations, which DKMS does not support). We could consider (stable) masking the flag for now, at least until I get the documentation situation in order.
- Yes this is a second path to support, but as the maintainer of the eclass, this maintenance load falls mostly on my shoulders. As already mentioned, this is an industry standard solution, often supported and tested upstream. Therefore once it is working I do not expect it to suddenly break on version bumps. This concern is valid, but note that I am intentionally **not** re-inventing the wheel here, just re-using what already works well on the Fedoras and the Debians etc.

Documentation is something that I will of course work on, but I'd like to finish the code part first. I already have some ideas to make things a bit prettier on the consumer end, which will mostly involve splitting some things out into smaller helper functions that can then be used in the ebuild.

Flowdalic commented on app-admin/ryzen_smu/ryzen_smu-0.1.5.ebuild:
```diff
@@ -15,16 +15,26 @@ KEYWORDS="amd64 x86"
 	doins "${FILESDIR}"/ryzen_smu.conf
 }
+
+pkg_postinst() {
```
Why are these added if the dkms.eclass exports this phase?

Flowdalic replied:
Oh, I see, because linux-mod-r1.elcass also exports it.

eli-schwartz replied:
No, pkgcheck is solved by correctly documenting the eclass.

This PR includes:

```
# @PROVIDES: linux-mod-r1
```

which means pkgcheck should not complain. The entire point of the pkgcheck complaint is that if an eclass doesn't ***provide*** another eclass, then you cannot just assume it's available because idk, it happens to inherit flag-o-matic to use append-flags deep inside one deprecated function.

hhoffstaette commented:
My belated 0.02€ as I was only now notified of this thread and took a look at at the changes to scap-driver (thanks for doing that already!). I welcome a general dkms eclass for all the right reasons (industry support, easier porting to Gentoo etc.), but did not expect an extension to linux-mod-r1 but rather expected it as an alternative, ie. using either dkms or inux-mod.
Is the intention to move away from linux-mod and eventually deprecate it?

hlein reviewed (approved):
I have zero interest in DKMS support but also zero objection.

Particularly for `loop-aes`, systems using it are likely to use it for `/`, meaning "recompile missing modules upon reboot" will definitely not work - if you're missing it you're not booting. But that doesn't mean adding DKMS support is in any way harmful (taking your word for it that `s/inherit linux-mod-r1/inherit dkms/` is effectively a noop when `USE=-dkms`).

Nowa-Ammerlaan commented:
I'm quite happy with how this looks now, particular the consumer end looks a lot nicer now. Please do let me know if you have any additional comments/concerns/objections.

Next step would be to write the documentation, the very brief summary for the user end is:
- enable `USE=dkms` and rebuild affected packages
- optionally customize/configure module signing/compression in `/etc/dkms/framework.conf`
- run `systemctl enable --now dkms`

And that's it. Kernel modules will then be (re)built automatically:
- if they are missing at boot
- by `kernel-install`
- by `emerge --config PN`

Nowa-Ammerlaan commented:
Added here some changes for `dist-kernel-utils.eclass` (CC @gentoo/dist-kernel ) to make use of `>=sys-kernel/installkernel-56`'s new capability of processing and passing optional arguments. This is technically independent of the proposed `dkms.eclass` (so can be merged separately) but it is thematically related because the new support for the `--all` argument is particularly useful in the  `dkms.eclass` to rebuild all installed distribution kernels when the `MODULES_OPTIONAL_IUSE` is enabled. 

If you weren't convinced yet that support for DKMS is worth the effort, then maybe this will help as a concrete example. `USE="dist-kernel initramfs dkms" emerge -1u zfs-kmod` will now trigger re-installation of the initramfs for **all** distribution kernels installed on the system and therefore (via the `dkms.install` `kernel-install` hook)  the update for the zfs kernel module is properly propagated on systems that have multiple distribution kernels installed in parallel. Of course `dist-kernel_install_kernel` retains compatibility with the use case of reinstalling a single kernel version via `linux-mod-r1.eclass`, as well as backwards compatibility with installkernel versions older then 56.

Nowa-Ammerlaan commented:
Work in progress documentation: https://wiki.gentoo.org/wiki/DKMS

Nowa-Ammerlaan commented:
Little fix to get DKMS to respect our `MODULES_SIGN_KEY` and `MODULES_SIGN_CERT`: https://github.com/dell/dkms/pull/498

It will be in the next release, and effectively avoids the need to setup module signing in two places. Which eliminates a major drawback of `dkms.eclass`.

mgorny commented on eclass/dist-kernel-utils.eclass:
```diff
@@ -66,18 +66,56 @@ dist-kernel_get_image_path() {
+	for arg in "${@}"; do
+		case "${arg}" in
+			-*) extrargs+=( "${arg}" );;
+			*)
+				let counter=counter+1
+				if [[ ${counter} -eq 1 ]]; then
+					version=${arg}
+				elif [[ ${counter} -eq 2 ]]; then
+					image=${arg}
+				elif [[ ${counter} -eq 3 ]]; then
+					map=${arg}
+				elif [[ ${counter} -eq 4 ]]; then
+					dir=${arg}
+				else
+					die "${FUNCNAME}: Too many arguments"
+				fi
+			;;
```
This logic is horribly complex, and on top of that you're changing the existing behavior in the same commit. Let's split it first, and then consider how to improve the logic.

Nowa-Ammerlaan replied:
Most of this complexity is because I was trying to support an arbitrary order of optional and positional arguments. I can simplify this considerably by forcing that the optional arguments must come last, which comes with the only drawback of reduced flexibility.

Once the new version of `installkernel` is stabilized we can drop the version condition and just pass on the `${@}` to `installkernel` directly. Which would make it even simpler.

Nowa-Ammerlaan replied:
> This logic is horribly complex, and on top of that you're changing the existing behavior in the same commit. Let's split it first, and then consider how to improve the logic.

How about now? This is significantly simpler, functionally equivalent, but comes with the drawback of less flexibility in terms of the ordering of the optional arguments versus the positional arguments, meaning they must now always come last.

mgorny commented on eclass/dist-kernel-utils.eclass:
```diff
@@ -66,18 +66,56 @@ dist-kernel_get_image_path() {
+		"${version}" "${image}" "${map}" "${dir}"
+	)
+	# Versions prior to 56 have no support for parsing optional args.
+	if has_version ">=sys-kernel/installkernel-56"; then
```
Is it actually safe to conditionally ignore additional flags?

Nowa-Ammerlaan replied:
Yes, because so far the only optional arguments supported are `--help`, `--verbose` (which used to be the default anyway), `--(no-)systemd`(to override the use of `kernel-install`), and `--all`. We could add more in the future, but for now `--all` is really the only one of any relevance to the eclass/ebuilds. For older versions of installkernel the logic remains unchanged, the specified kernel version is reinstalled and `--all` is ignored.

thesamesam commented on eclass/dkms.eclass:
```diff
@@ -0,0 +1,527 @@
+
+	while IFS= read -r -d '' file; do
+		dkms_dopackage $(dirname "${file}")
+	done < <(find . -type f -name dkms.conf -print0)
```
Missing die?

thesamesam commented on eclass/dkms.eclass:
```diff
@@ -0,0 +1,527 @@
+		fi
+
+		cat <<-EOF >> dkms.conf || die
+
```
The new line at the start looks off. If it's intentional, add a comment header instead?

Nowa-Ammerlaan replied:
It's intentional, I encountered one instance of an upstream stub `dkms.conf` that was missing a newline at the end of the file which lead to a mess when trying to append to it. The extra new line prevents a mess in this case. I'll add a comment.

thesamesam commented on eclass/dkms.eclass:
```diff
@@ -0,0 +1,527 @@
+		[[ -f ${file} ]] ||
+			die "${FUNCNAME}: DKMS conf does not exist: ${file}"
+		# Check if make command already respecting our options.
+		# If not, sed our options in for all MAKE[#].
```
This is a bit gnarly. Is this a bug in dkms, or what? Or just people writing crappy Makefiles and we want to fix them up to work with dkms? I don't know we should have this workaround.

Nowa-Ammerlaan replied:
It's just a check to prevent sed'ing in the `MODULES_MAKEARGS` more then once since this function is called for each module in the `modlist`.

thesamesam replied:
Oh, I see. But if it's not respecting it, why is this our problem? Why are we mangling it in the eclass at all?

Nowa-Ammerlaan replied:
> Perhaps I could rewrite this a bit to take this part of the `if` outside of the `for` loop.

Actually no, we cannot do that since the `modlist` may contain modules in different source directories (`mod[1]`) and then we may create multiple `dkms.conf`s, each would have to be `dkms_gentoofy_conf`'ed if it exists already but if there are duplicate source dirs in the `modlist` then we do still need this check to verify that we have not already run `dkms_gentoofy_conf` on that `dkms.conf`.

Nowa-Ammerlaan replied:
> This is a bit gnarly.

Restructured this a bit. Now we keep track of the `dkms.conf` files we touch, and then afterwards we call `dkms_gentoofy_conf` them outside of the `for` loop. This means we can now safely assume this function is only called once and we can therefore drop this weird condition. As an additional bonus, this also means we only have to call `dkms_sanitize_makeargs` in one place, which makes the structure as a whole look a bit cleaner. Let me know what you think!

Also rebased the PR.

thesamesam commented on eclass/dkms.eclass:
```diff
@@ -0,0 +1,527 @@
+if [[ -z ${_DKMS_ECLASS} ]]; then
+_DKMS_ECLASS=1
+
+inherit linux-mod-r1
```
I appreciate that this is the natural solution if Ionen isn't sure (or rejects it) for linux-mod-r1, but having basically every ebuild change from `linux-mod-r1` to `kmod` for what it inherits instead feels like essentially the same thing.

Nowa-Ammerlaan replied:
> instead feels like essentially the same thing.

I suppose. But a fundamental difference with my first attempt is that now we're not adding `USE=dkms` to everything blindly, which was a big flaw in my original design where this was all part of `linux-mod-r1`. I actually tested this now on each package where the new flag is added and verified that it works.

Nowa-Ammerlaan commented on eclass/dist-kernel-utils.eclass:
```diff
@@ -66,18 +66,35 @@ dist-kernel_get_image_path() {
+# an overview of which optional arguments are supported. Of particular
+# interest is the --all argument which may be used to iteratively
+# (re-)install all distribution kernels installed on the system.
+# Optional arguments must come after the positional arguments.
```
Done! This does look a lot cleaner.

mgorny commented on eclass/dist-kernel-utils.eclass:
```diff
@@ -66,18 +66,38 @@ dist-kernel_get_image_path() {
+#
+# <dir> -- the target directory for install (default: ${EROOT}/boot)
+#
+# <installkernel-args> -- extra optional arguments for installkernel
+# 		see man kernel-install and man installkernel. Requires at
+# 		least version 56 of sys-kernel/installkernel. Optional
+# 		arguments must come after the positional arguments.
```
```suggestion
# <installkernel-args> -- additional arguments to pass to installkernel,
# 		see man kernel-install and man installkernel. Require
# 		>=sys-kernel/installkernel-56.
```

The order is implied by usage, and talking about it here suggests that there are "positional" and "optional" arguments to installkernel.

Nowa-Ammerlaan commented on eclass/dist-kernel-utils.eclass:
```diff
@@ -66,18 +66,38 @@ dist-kernel_get_image_path() {
+# <system.map> -- the full path to the System.map (default:
+# 		${KV_DIR}/System.map})
+#
+# <dir> -- the target directory for install (default: ${EROOT}/boot)
```
Done!

eli-schwartz commented:
IMHO there should be a way to override this per package to copy something other than ${S}, and then packages which build other things in addition to a kernel module would be able to filter out the sources and tell dkms.eclass to use that instead.

Nowa-Ammerlaan commented:
> IMHO there should be a way to override this per package to copy something other than ${S}, and then packages which build other things in addition to a kernel module would be able to filter out the sources and tell dkms.eclass to use that instead.

This is already the case, the `modlist` effectively customizes how much of the sources are included.

> If I understand this correctly the eclass will copy S somewhere to store the sources for later. In this case using dkms for app-emulation/virtualbox-guest-additions is wasteful,

This intentionally does *not* install the whole of `S`, we use the `modlist` from `linux-mod-r1` which defines the location of the kernel module sources. The `dkms.conf` is created there, and not in `S`. 

In the install phase we look for the `dkms.conf` file(s) and install directories containing those files. End result is that we only install the kernel module sources.

For the case of `virtualbox-guest-additions` this means that the `VBOX_MOD_SRC_DIR` is installed, which is only `5.8M`:

```
nowa-gentoo-pc src # du -h virtualbox-guest-additions_vboxguest-7.1.6
44K     virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/VBox
92K     virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/common/misc
44K     virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/common/err
60K     virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/common/time
324K    virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/common/string
52K     virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/common/table
40K     virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/common/alloc
8.0K    virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/common/asm
16K     virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/common/checksum
44K     virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/common/math/gcc
48K     virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/common/math
196K    virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/common/log
884K    virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/common
24K     virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/include/VBox/HostServices
12K     virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/include/VBox/vmm
560K    virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/include/VBox
16K     virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/include/iprt/linux
8.0K    virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/include/iprt/nocrt
2.7M    virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/include/iprt
112K    virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/include/internal
3.4M    virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/include
56K     virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/generic
24K     virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/r0drv/generic
324K    virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/r0drv/linux
448K    virtualbox-guest-additions_vboxguest-7.1.6/vboxguest/r0drv
5.2M    virtualbox-guest-additions_vboxguest-7.1.6/vboxguest
4.0K    virtualbox-guest-additions_vboxguest-7.1.6/vboxsf/VBox
144K    virtualbox-guest-additions_vboxguest-7.1.6/vboxsf/include/VBox
4.0K    virtualbox-guest-additions_vboxguest-7.1.6/vboxsf/include/iprt/linux
4.0K    virtualbox-guest-additions_vboxguest-7.1.6/vboxsf/include/iprt/nocrt
36K     virtualbox-guest-additions_vboxguest-7.1.6/vboxsf/include/iprt
184K    virtualbox-guest-additions_vboxguest-7.1.6/vboxsf/include
4.0K    virtualbox-guest-additions_vboxguest-7.1.6/vboxsf/r0drv/linux
8.0K    virtualbox-guest-additions_vboxguest-7.1.6/vboxsf/r0drv
624K    virtualbox-guest-additions_vboxguest-7.1.6/vboxsf
5.8M    virtualbox-guest-additions_vboxguest-7.1.6
```

orlitzky commented:
tl;dr but ACK for tp_smapi

hhoffstaette commented:
Wearing my scap-driver hat the latest version looks more like what I originally expected, so :+1:

jayofdoom commented:
I suspect there's a lot here I don't understand, but I have a question-maybe-concern about this:

It seems to me, one of the two possibilities below are true; I'm not sure which one but neither is pleasant.
1. It's now possible that if I set USE=dkms, I might have TWO ways modules build on my system if I have one that uses linux-mod-r1 directly and one that uses dkms **OR**
2. It's now mandatory for all modules in-tree to use the new dkms eclass (or at least will be once stable); making it weird to me that it's not just made part of linux-mod-r1 directly.

As a user; the first possibility is extremely unpleasant. I now have to learn how to troubleshoot two different things when a module is missing. Plus what happens with $mod_no_dkms is missing from an old kernel, but $dkms_mod isn't -- it's going to be even more confusing.

As a developer, I don't like the churn of the way it's being done now as a separate eclass; so if the second possibility (all modules must use dkms eclass); then why make it as a new eclass instead of modifying the existing. Unless the plan would be to deprecate linux-mod-r1 eventually and supercede it, this just seems like we're setting people up for failure by essentially leaving the old way enabled.

What am I missing?

eli-schwartz commented:
I'm not sure I understand the concern for option 1.

If you don't want a mix of methods then just don't use the USE flag. The USE flag is still useful for people with a system where all used modules are able to handle dkms.

There is a little bit more end-user complexity around kernel installation and usage but that complexity can be concisely summarized as "decide whether it's worth bothering with dkms, and if in doubt, continue as you have in the past".

I'd also like point out at this time, that coming from other distros with frankly a lot less tolerance for choice than Gentoo has, it was still offered as a choice, to take your pick of binpkg'ed precompiled modules or dkms sources.

Nobody ever cried about having a choice, if I may be blunt. Some people only ever installed "module" packages, and some people only ever installed "dkms" packages, and both sides were happy and neither side ever got confused.

The only thing that confuses me is how Gentoo users could end up being more confused by choice than Arch Linux users.

Nowa-Ammerlaan commented:
> It seems to me, one of the two possibilities below are true; I'm not sure which one but neither is pleasant.
> 
>     1. It's now possible that if I set USE=dkms, I might have TWO ways modules build on my system if I have one that uses linux-mod-r1 directly and one that uses dkms **OR**

I don't think it makes sense to toggle this flag on a per package basis. Similar to for example the `dist-kernel` flag, this `dkms` flag is intended to be either enabled or disabled globally.  As @eli-schwartz already pointed out above, either the user chooses to use DKMS or they do not. 

>     2. It's now mandatory for all modules in-tree to use the new dkms eclass (or at least will be once stable); making it weird to me that it's not just made part of linux-mod-r1 directly.

Nothing is mandatory. As I've said before, I am not suggesting to replace anything. I honestly do not understand why people keep thinking that this is what I am suggesting, I never said this. `linux-mod-r1.eclass` is great, all I am doing here is adding a second option that makes new things possible that we cannot achieve with the package manager. In the end it is up to the ebuild maintainer to decided to support dkms or not, and then it is up to the user to decide to use it or not. That being said, for a coherent user experience I do think it would be best to add support for it everywhere where it is easy and cheap to do so (as I have already done in this PR). It is precisely for this reason that I went through the trouble of writing a function to convert an ebuilds `modlist` to a `dkms.conf`, this makes it super easy for the ebuild maintainer to add support for dkms to an ebuild.

jayofdoom commented:
> If you don't want a mix of methods then just don't use the USE flag. The USE flag is still useful for people with a system where all used modules are able to handle dkms.

In a world where it seems "free" to add dkms to a kernel moodule, I guess I don't see the virtue in /not/ having it enabled everywhere. Is there a technical piece here I'm missing.

>  As I've said before, I am not suggesting to replace anything. I honestly do not understand why people keep thinking that this is what I am suggesting, I never said this.

I gleaned this as a likely possibility from the fact multiple packages (including the one I maintain; thank you!) already being updated. The crux of my question is trying to discover the desired end state -- my assumption was (maybe falsely?) that this was proposing a middle state for migration that might not represent the end state once completed.

eli-schwartz commented:
> > If you don't want a mix of methods then just don't use the USE flag. The USE flag is still useful for people with a system where all used modules are able to handle dkms.
> 
> In a world where it seems "free" to add dkms to a kernel moodule, I guess I don't see the virtue in /not/ having it enabled everywhere. Is there a technical piece here I'm missing?
> 
> edit:
> 
> > The only thing that confuses me is how Gentoo users could end up being more confused by choice than Arch Linux users.
> 
> To be clear: I like the fact it's a use flag users can toggle. I don't love that if I put USE=dkms in my /etc/portage/make.conf that I'm not guaranteed all my modules will use dkms. I understand if it's not possible; but I don't understand why it's controversial to say it's non-ideal to have different behavior.

My answer to that would be as I said before:

Putting USE=dkms is a personal choice users can do if they know all their modules support dkms. If not all their modules support dkms, then they will have to decide whether it's a pain for them to have different behavior.

It's literally just... choices. Additional choices means a bit more work in documentation, to explain the tradeoffs surrounding how users should make the choice (of whether or not to put USE=dkms in their make.conf), and I don't think that additional documentation alongside additional choices is an insurmountable barrier to Gentoo users. Or any barrier at all, really.

If someone is worried about "non-ideal different behavior" then obviously they fall into the camp of people for whom USE=dkms does not align with their needs, and they simply don't use it.

> The crux of my question is trying to discover the desired end state -- my assumption was (maybe falsely?) that this was proposing a middle state for migration that might not represent the end state once completed.

My assumption regarding the desired end state is:

- both built modules and dkms sources are valid methods for different users, both provide features that the other does not and *cannot* handle, and both are worth supporting
- in principle, there's really no reason at all that any package would be unable to support both
  - built modules can literally be built *by using dkms to build them*, I can directly testify to this since I personally revolutionized kernel modules on Arch Linux by doing precisely this. For background, Arch has split packages but if you build binpackages for both built modules and dkms, there is no such thing as slot dependencies so you must bump the revision to a new `-rXX` every single time a new kernel is released, leading to dkms packages that look like vhba-module-dkms-20240917-r33 and constantly update once a week. Building each as a *separate* package means maintaining sources and *patches* separately. Instead, you can have the other package in BDEPEND and run `dkms build --dkmstree "${S}" -m ${PN}/${PV} -k ${KV}` and it simply... works. Regardless of what dkms-specific features are used, including all sorts of conditional patching.
  - dkms packages are simply a config file that runs the equivalent of `src_configure && src_compile` and it's not impractical to write them.


So we would ultimately want to support both for all packages and then people wouldn't have to worry about whether USE=dkms produces a mixed state. But each package has to be ported. There's a transition period here, and the transition period is simply "not all packages support user choice". And it's also up to package maintainers, whether they are comfortable implementing it especially e.g. testing to make sure it is correctly configured.

That might mean it's never "complete" as some packages never end up supporting dkms, but this isn't a reflection on the desired end state, and if there are packages that don't support it then we can ultimately just document that on the wiki as "if you use any of these modules, don't bother with USE=dkms as you won't be able to use it globally anyway".

jayofdoom commented:
Thanks for providing some clarity as to the larger idea. That's all I wanted to understand :D

Nowa-Ammerlaan commented:
Loosely related: https://github.com/gentoo/gentoo/pull/40943 (to make it easier to simultaneously install different `major.minor` versions of the kernel)

Nowa-Ammerlaan commented:
Since I have so far received only one comment on my RFC, which was positive. I propose to move this forward. I have rebased this PR and will send the eclass patches to the mailing list. Barring any objections I propose to merge late next week or the week after that.

hhoffstaette commented:
@Nowa-Ammerlaan https://github.com/Nowa-Ammerlaan/gentoo/pull/1 :)

Nowa-Ammerlaan commented:
Merged into https://github.com/nowa-Ammerlaan/natinst for now
